planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,438 @@
+      if (firstNode == null) {
+        firstNode = new Node();
+        nodes.put(first, firstNode);
+        output.add(firstNode);
+      }
+
+      Coordinate last = coordinateSequence.getLast();
+      Node lastNode = nodes.get(last);
+      if (lastNode == null) {
+        lastNode = new Node();
+        nodes.put(last, lastNode);
+        output.add(lastNode);
+      }
+
+      double length = length(coordinateSequence);
+
+      Edge edge = new Edge(firstNode, lastNode, coordinateSequence, length);
+
+      firstNode.addEdge(edge);
+      lastNode.addEdge(edge.reversed);
+    }
+  }
+
+  private List<List<Coordinate>> nodeLines(List<LineString> input) {
+    Map<Coordinate, Integer> nodeCounts = new HashMap<>();
+    List<List<Coordinate>> coords = new ArrayList<>(input.size());
+    for (var line : input) {
+      var coordinateSequence = line.getCoordinateSequence();
+      List<Coordinate> snapped = new ArrayList<>();
+      Coordinate last = null;
+      for (int i = 0; i < coordinateSequence.size(); i++) {
+        Coordinate current = new CoordinateXY(coordinateSequence.getX(i), coordinateSequence.getY(i));
+        precisionModel.makePrecise(current);
+        if (last == null || !last.equals(current)) {
+          snapped.add(current);
+          nodeCounts.merge(current, 1, Integer::sum);
+        }
+        last = current;
+      }
+      if (snapped.size() >= 2) {
+        coords.add(snapped);
+      }
+    }
+
+    List<List<Coordinate>> result = new ArrayList<>(input.size());
+    for (var coordinateSequence : coords) {
+      int start = 0;
+      for (int i = 0; i < coordinateSequence.size(); i++) {
+        Coordinate coordinate = coordinateSequence.get(i);
+        if (i > 0 && i < coordinateSequence.size() - 1 && nodeCounts.get(coordinate) > 1) {
+          result.add(coordinateSequence.subList(start, i + 1));
+          start = i;
+        }
+      }
+      if (start < coordinateSequence.size()) {
+        var sublist = start == 0 ? coordinateSequence : coordinateSequence.subList(start, coordinateSequence.size());
+        result.add(sublist);
+      }
+    }
+    return result;
+  }
+
+  class Node {
+    int id = nodes++;
+    List<Edge> edge = new ArrayList<>();
+
+    void addEdge(Edge edge) {
+      for (Edge other : this.edge) {
+        if (other.coordinates.equals(edge.coordinates)) {
+          return;
+        }
+      }
+      this.edge.add(edge);
+    }
+
+    List<Edge> getEdges() {
+      return edge;
+    }
+
+    void removeEdge(Edge edge) {
+      this.edge.remove(edge);
+    }
+
+    @Override
+    public String toString() {
+      return "Node{" + id + ": " + edge + '}';
+    }
+
+    public double distance(Node end) {
+      if (!getEdges().isEmpty() && !end.getEdges().isEmpty()) {
+        Coordinate a = getEdges().getFirst().coordinates.getFirst();
+        Coordinate b = end.getEdges().getFirst().coordinates.getFirst();
+        return a.distance(b);
+      }
```

Comment:
it might make sense to just put coordinate on each node so you don't need to look at the edges to get its location.